File: rust/benches/cast_benchmarks.rs
//! Criterion benchmarks pitting each door against the closest established parser — the
//! stdlib's own `str::parse` for booleans/numerics, the `uuid` crate for UUID text, and the
//! `time` crate's RFC 3339 well-known format for instants. Numbers are informational this
//! round; the load-bearing claims (allocation-free, corpus conformance) live in the tests.

use criterion::{criterion_group, criterion_main, Criterion};
use hypercast::{DateOrder, NumFormat};
use std::hint::black_box;
use time::format_description::well_known::Rfc3339;

fn benchmarks(c: &mut Criterion) {
    let format = NumFormat::INVARIANT;

    let mut group = c.benchmark_group("bool");
    group.bench_function("hypercast", |b| {
        b.iter(|| hypercast::cast_bool(black_box(b"true")).unwrap())
    });
    group.bench_function("stdlib", |b| {
        b.iter(|| black_box("true").parse::<bool>().unwrap())
    });
    group.finish();

    let mut group = c.benchmark_group("i64");
    group.bench_function("hypercast", |b| {
        b.iter(|| hypercast::cast_i64(black_box(b"1234567890123"), &format).unwrap())
    });
    group.bench_function("stdlib", |b| {
        b.iter(|| black_box("1234567890123").parse::<i64>().unwrap())
    });
    group.finish();

    let mut group = c.benchmark_group("f64");
    group.bench_function("hypercast", |b| {
        b.iter(|| hypercast::cast_f64(black_box(b"12345.6789"), &format).unwrap())
    });
    group.bench_function("stdlib", |b| {
        b.iter(|| black_box("12345.6789").parse::<f64>().unwrap())
    });
    group.finish();

    let mut group = c.benchmark_group("uuid");
    group.bench_function("hypercast", |b| {
        b.iter(|| hypercast::cast_uuid(black_box(b"01020304-0506-0708-090a-0b0c0d0e0f10")).unwrap())
    });
    group.bench_function("uuid-crate", |b| {
        b.iter(|| uuid::Uuid::try_parse(black_box("01020304-0506-0708-090a-0b0c0d0e0f10")).unwrap())
    });
    group.finish();

    let mut group = c.benchmark_group("timestamp");
    group.bench_function("hypercast", |b| {
        b.iter(|| hypercast::cast_timestamp(black_box(b"2026-01-02T15:04:05.123456789Z")).unwrap())
    });
    group.bench_function("time-crate", |b| {
        b.iter(|| {
            time::OffsetDateTime::parse(black_box("2026-01-02T15:04:05.123456789Z"), &Rfc3339)
                .unwrap()
        })
    });
    group.finish();

    // The messy civil shape: no Rust stdlib parser accepts it at all, so the pairing is
    // against the `time` crate driven by a hand-built format description — the closest
    // equivalent work, same as the Java/C# tables pair against pattern formatters.
    let mut group = c.benchmark_group("datetime");
    group.bench_function("hypercast-messy", |b| {
        b.iter(|| {
            hypercast::cast_datetime(black_box(b"1/7/2026 3:04 PM"), DateOrder::MonthDayYear)
                .unwrap()
        })
    });
    group.bench_function("hypercast-iso", |b| {
        b.iter(|| {
            hypercast::cast_datetime(black_box(b"2026-01-07T15:04:05"), DateOrder::YearMonthDay)
                .unwrap()
        })
    });
    group.bench_function("hypercast-date-ordered", |b| {
        b.iter(|| {
            hypercast::cast_date_ordered(black_box(b"1/7/2026"), DateOrder::MonthDayYear).unwrap()
        })
    });
    group.finish();

    // Separator detection against the same text under a declared format: the cost of the
    // structural resolution pass, isolated.
    let mut group = c.benchmark_group("separator-detection");
    const EUROZONE: NumFormat =
        NumFormat { decimal_sep: ',', group_sep: '.', flags: NumFormat::ALL };
    group.bench_function("hypercast-detect", |b| {
        b.iter(|| hypercast::cast_f64(black_box(b"1.234.567,89"), &NumFormat::DETECT).unwrap())
    });
    group.bench_function("hypercast-declared", |b| {
        b.iter(|| hypercast::cast_f64(black_box(b"1.234.567,89"), &EUROZONE).unwrap())
    });
    group.finish();

    let mut group = c.benchmark_group("duration");
    group.bench_function("hypercast-iso", |b| {
        b.iter(|| hypercast::cast_duration(black_box(b"P1DT6H30M15.5S")).unwrap())
    });
    group.bench_function("hypercast-colon", |b| {
        b.iter(|| hypercast::cast_duration(black_box(b"1.06:30:15.5")).unwrap())
    });
    group.finish();
}

criterion_group!(benches, benchmarks);
criterion_main!(benches);

File: java/benchmarks/src/jmh/java/io/github/skunkwerkx/hypercast/benchmarks/CastBenchmarks.java
package io.github.skunkwerkx.hypercast.benchmarks;

import io.github.skunkwerkx.hypercast.Cast;
import io.github.skunkwerkx.hypercast.DateOrder;
import io.github.skunkwerkx.hypercast.NumFormat;
import io.github.skunkwerkx.hypercast.Verdict;
import java.lang.foreign.MemorySegment;
import java.nio.charset.StandardCharsets;
import java.text.NumberFormat;
import java.text.ParseException;
import java.time.Duration;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;

/**
 * Each pair pits a HyperCast door (FFM crossing included) against the closest JDK parse
 * with equivalent settings declared — {@link NumberFormat} where grouping needs the JDK's
 * locale machinery, {@code java.time} parsers where that's simply how the JDK door works.
 * String-door numbers include the UTF-8 encode; the {@code byte[]} rows show the raw
 * crossing.
 */
@State(Scope.Benchmark)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class CastBenchmarks {
    private static final NumFormat INVARIANT = NumFormat.INVARIANT;
    private static final NumberFormat JDK_GROUPED = NumberFormat.getIntegerInstance(Locale.US);
    private static final DateTimeFormatter ISO_OFFSET = DateTimeFormatter.ISO_OFFSET_DATE_TIME;

    private final String boolText = "true";
    private final String intText = "1,234,567";
    private final String doubleText = "12345.6789";
    private final String uuidText = "01020304-0506-0708-090a-0b0c0d0e0f10";
    private final String timestampText = "2026-01-02T15:04:05.123456789Z";
    private final String offsetTimestampText = "2026-01-02T15:04:05.123456789+05:00";
    private final String isoDurationText = "PT1H30M15.5S";
    private final String timeText = "15:04:05.123456789";
    private final byte[] timestampBytes = timestampText.getBytes(StandardCharsets.UTF_8);
    private final byte[] intBytes = intText.getBytes(StandardCharsets.UTF_8);
    private final byte[] uuidBytes = uuidText.getBytes(StandardCharsets.UTF_8);
    // The round-three shape: one buffer holding many values, each door handed a heap slice.
    private final MemorySegment line = MemorySegment.ofArray(
            (intText + "|" + timestampText).getBytes(StandardCharsets.UTF_8));
    private final MemorySegment intSlice = line.asSlice(0, intBytes.length);
    private final MemorySegment timestampSlice = line.asSlice(intBytes.length + 1, timestampBytes.length);

    @Benchmark
    public Verdict<Boolean> castBool() {
        return Cast.bool(boolText);
    }

    @Benchmark
    public boolean jdkParseBoolean() {
        // Never fails by design — anything not "true" is silently false. Not equivalent
        // machinery, printed for scale only.
        return Boolean.parseBoolean(boolText);
    }

    @Benchmark
    public Verdict<Integer> castI32Grouped() {
        return Cast.i32(intText, INVARIANT);
    }

    @Benchmark
    public Verdict<Integer> castI32GroupedBytes() {
        return Cast.i32(intBytes, INVARIANT);
    }

    @Benchmark
    public Verdict<Integer> castI32GroupedSlice() {
        return Cast.i32(intSlice, INVARIANT);
    }

    @Benchmark
    public Number jdkNumberFormatGrouped() throws ParseException {
        // The JDK's own grouped-integer door. Stateful and not thread-safe, unlike Cast.
        return JDK_GROUPED.parse(intText);
    }

    @Benchmark
    public Verdict<Double> castF64() {
        return Cast.f64(doubleText, INVARIANT);
    }

    @Benchmark
    public double jdkParseDouble() {
        return Double.parseDouble(doubleText);
    }

    @Benchmark
    public Verdict<UUID> castUuid() {
        return Cast.uuid(uuidText);
    }

    @Benchmark
    public Verdict<UUID> castUuidBytes() {
        return Cast.uuid(uuidBytes);
    }

    @Benchmark
    public UUID jdkUuidFromString() {
        return UUID.fromString(uuidText);
    }

    @Benchmark
    public Verdict<Instant> castTimestamp() {
        return Cast.timestamp(timestampText);
    }

    @Benchmark
    public Verdict<Instant> castTimestampBytes() {
        return Cast.timestamp(timestampBytes);
    }

    @Benchmark
    public Verdict<Instant> castTimestampSlice() {
        return Cast.timestamp(timestampSlice);
    }

    @Benchmark
    public Instant jdkInstantParse() {
        return Instant.parse(timestampText);
    }

    @Benchmark
    public Verdict<Instant> castTimestampOffset() {
        return Cast.timestamp(offsetTimestampText);
    }

    @Benchmark
    public Instant jdkOffsetDateTimeParse() {
        return OffsetDateTime.parse(offsetTimestampText, ISO_OFFSET).toInstant();
    }

    @Benchmark
    public Verdict<Duration> castDurationIso() {
        return Cast.duration(isoDurationText);
    }

    @Benchmark
    public Duration jdkDurationParse() {
        return Duration.parse(isoDurationText);
    }

    @Benchmark
    public Verdict<LocalTime> castTime() {
        return Cast.time(timeText);
    }

    @Benchmark
    public LocalTime jdkLocalTimeParse() {
        return LocalTime.parse(timeText);
    }
    // --- the declared-order doors, vs the JDK pattern formatter that could accept the
    // same text. This is the messy-feed shape the doors exist for: "1/7/2026 3:04 PM"
    // has no java.time parser of its own, only a hand-built formatter.

    private static final DateTimeFormatter US_DATE_TIME =
            DateTimeFormatter.ofPattern("M/d/yyyy h:mm a", Locale.US);
    private static final DateTimeFormatter US_DATE = DateTimeFormatter.ofPattern("M/d/yyyy", Locale.US);

    private final String messyDateTimeText = "1/7/2026 3:04 PM";
    private final String isoDateTimeText = "2026-01-07T15:04:05";
    private final String messyDateText = "1/7/2026";

    @Benchmark
    public Verdict<LocalDateTime> castDateTimeMessy() {
        return Cast.dateTime(messyDateTimeText, DateOrder.MONTH_DAY_YEAR);
    }

    @Benchmark
    public LocalDateTime jdkPatternDateTimeParse() {
        return LocalDateTime.parse(messyDateTimeText, US_DATE_TIME);
    }

    @Benchmark
    public Verdict<LocalDateTime> castDateTimeIso() {
        return Cast.dateTime(isoDateTimeText, DateOrder.YEAR_MONTH_DAY);
    }

    @Benchmark
    public LocalDateTime jdkLocalDateTimeParse() {
        return LocalDateTime.parse(isoDateTimeText);
    }

    @Benchmark
    public Verdict<LocalDate> castDateOrdered() {
        return Cast.date(messyDateText, DateOrder.MONTH_DAY_YEAR);
    }

    @Benchmark
    public LocalDate jdkPatternDateParse() {
        return LocalDate.parse(messyDateText, US_DATE);
    }

    // --- separator detection, vs the same text under a declared eurozone format and vs
    // the JDK's own locale machinery. Detection's cost is one extra scan for '.'/','.

    private static final NumFormat EUROZONE =
            new NumFormat(',', '.', NumFormat.STYLE_ALL);
    private static final NumberFormat JDK_GERMAN = NumberFormat.getInstance(Locale.GERMANY);

    private final String euroNumberText = "1.234.567,89";

    @Benchmark
    public Verdict<Double> castF64Detect() {
        return Cast.f64(euroNumberText, NumFormat.DETECT);
    }

    @Benchmark
    public Verdict<Double> castF64Declared() {
        return Cast.f64(euroNumberText, EUROZONE);
    }

    @Benchmark
    public Number jdkNumberFormatGerman() throws ParseException {
        return JDK_GERMAN.parse(euroNumberText);
    }

}
